## Northwind Traders Analytics

#### Setup and Load PostgreSQL database
- Connect to PostgreSQL database and load SQL extension
- Perform data exploration to view all base tables
- Create tables with common joins for querying throughout the rest of the project

In [ ]:
import pandas as pd
import matplotlib as plt
import seaborn as sns

In [4]:
%load_ext sql
%sql postgresql://postgres@localhost:5432/northwind

In [5]:
%%sql

SELECT table_name, table_type 
  FROM information_schema.tables 
 WHERE table_schema = 'public';

 * postgresql://postgres@localhost:5432/northwind
17 rows affected.


[('territories', 'BASE TABLE'),
 ('order_details', 'BASE TABLE'),
 ('employee_territories', 'BASE TABLE'),
 ('us_states', 'BASE TABLE'),
 ('customers', 'BASE TABLE'),
 ('orders', 'BASE TABLE'),
 ('employees', 'BASE TABLE'),
 ('shippers', 'BASE TABLE'),
 ('products', 'BASE TABLE'),
 ('categories', 'BASE TABLE'),
 ('suppliers', 'BASE TABLE'),
 ('region', 'BASE TABLE'),
 ('customer_demographics', 'BASE TABLE'),
 ('customer_customer_demo', 'BASE TABLE'),
 ('customer_orders', 'VIEW'),
 ('product_quantity_orders', 'VIEW'),
 ('employee_orders', 'VIEW')]

In [119]:
%%sql 

CREATE OR REPLACE VIEW customer_orders AS 
SELECT c.*, o.order_id, o.employee_id, o.order_date, o.required_date, o.shipped_date, o.ship_via, o.freight, o.ship_name, o.ship_address, o.ship_city, o.ship_region, o.ship_postal_code, o.ship_country
  FROM customers AS c
  JOIN orders AS o 
    ON c.customer_id = o.customer_id;

CREATE OR REPLACE VIEW product_quantity_orders AS
SELECT p.*, od.quantity, od.discount, o.*
  FROM products AS p
  JOIN order_details AS od ON p.product_id = od.product_id
  JOIN orders AS o ON od.order_id = o.order_id
 ORDER BY p.product_id;

CREATE OR REPLACE VIEW employee_orders AS 
SELECT e.last_name, e.first_name, e.title, e.title_of_courtesy, e.birth_date, e.hire_date, address, city, region, postal_code, country, home_phone, extension, notes, reports_to, photo_path, o.*
  FROM employees AS e
  JOIN orders AS o ON e.employee_id = o.employee_id
 ORDER BY o.employee_id;

 * postgresql://postgres@localhost:5432/northwind
Done.
Done.
Done.


[]

#### Employees Sales Performance

- Rank employees based on their total sales amount
- Compare each employee's sales performance to the average sales performance of all employees to evaluate their contributions.

In [145]:
%%sql

WITH aggregate_sales AS(
SELECT e.employee_id, SUM(od.unit_price * od.quantity * (1 - od.discount)) AS total_sales
  FROM employees AS e
  JOIN orders AS o ON e.employee_id = o.employee_id
  JOIN order_details AS od ON o.order_id = od.order_id
 GROUP BY e.employee_id)

SELECT employee_id, ROUND(total_sales::numeric,2) AS total_sales, RANK() OVER(ORDER BY total_sales DESC) AS employee_rank,
       ROUND(total_sales::numeric - (SELECT AVG(total_sales)::numeric FROM aggregate_sales),2) AS diff_from_average
  FROM aggregate_sales;

 * postgresql://postgres@localhost:5432/northwind
9 rows affected.


[(4, Decimal('232890.85'), 1, Decimal('92247.18')),
 (3, Decimal('202812.84'), 2, Decimal('62169.17')),
 (1, Decimal('192107.60'), 3, Decimal('51463.93')),
 (2, Decimal('166537.76'), 4, Decimal('25894.08')),
 (8, Decimal('126862.28'), 5, Decimal('-13781.39')),
 (7, Decimal('124568.23'), 6, Decimal('-16075.44')),
 (9, Decimal('77308.07'), 7, Decimal('-63335.60')),
 (6, Decimal('73913.13'), 8, Decimal('-66730.54')),
 (5, Decimal('68792.28'), 9, Decimal('-71851.39'))]

High performers are employees 4,3,1 and low performers are employees 9,6,5

#### Monthly Sales Trends

- Visualize the progress of the sales and identify trends that might shape the company's future strategies
- Aggregate sales data at a monthly level and calculating a running total of sales by month
- Analyze the month-over-month sales growth rate

In [147]:
%%sql

WITH monthly_sales AS (
SELECT DATE_TRUNC('month', order_date) AS order_month, SUM(unit_price * quantity * (1 - discount)) AS monthly_sales
  FROM orders 
  JOIN order_Details ON orders.order_id = order_details.order_id
 GROUP BY order_month
 ORDER BY order_month ASC)

SELECT order_month, ROUND(monthly_sales::numeric,2) AS monthly_sales, 
       ROUND(LAG(monthly_sales) OVER w::numeric,2) AS previous_month_sales,
       ROUND((monthly_sales::numeric -  LAG(monthly_sales) OVER w)::numeric / LAG(monthly_sales) OVER w::numeric * 100,2) AS MoM_growth_rate,
       ROUND(SUM(monthly_sales) OVER w::numeric,2) AS running_total
  FROM monthly_sales
WINDOW w AS (ORDER BY order_month);

 * postgresql://postgres@localhost:5432/northwind
23 rows affected.


[(datetime.datetime(1996, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('27861.90'), None, None, Decimal('27861.90')),
 (datetime.datetime(1996, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('25485.28'), Decimal('27861.90'), Decimal('-8.53'), Decimal('53347.17')),
 (datetime.datetime(1996, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('26381.40'), Decimal('25485.28'), Decimal('3.52'), Decimal('79728.57')),
 (datetime.datetime(1996, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('37515.72'), Decimal('26381.40'), Decimal('42.21'), Decimal('117244.30')),
 (datetime.datetime(1996, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('45600.05'), Decimal('37515.72'), Decimal('21.55'), Decimal('162844.34')),
 (datetime.datetime(1996, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('45239.63'), Decimal('45600.05'), Decimal('-0.79'), Decimal('208083.97')),
 (datetime.datetime(1997, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('61258.07'), Decimal('45239.63'), Decimal('35.41'), Decimal('269342.04')),
 (datetime.datetime(1997, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('38483.63'), Decimal('61258.07'), Decimal('-37.18'), Decimal('307825.68')),
 (datetime.datetime(1997, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('38547.22'), Decimal('38483.63'), Decimal('0.17'), Decimal('346372.90')),
 (datetime.datetime(1997, 4, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('53032.95'), Decimal('38547.22'), Decimal('37.58'), Decimal('399405.85')),
 (datetime.datetime(1997, 5, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('53781.29'), Decimal('53032.95'), Decimal('1.41'), Decimal('453187.14')),
 (datetime.datetime(1997, 6, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('36362.80'), Decimal('53781.29'), Decimal('-32.39'), Decimal('489549.94')),
 (datetime.datetime(1997, 7, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('51020.86'), Decimal('36362.80'), Decimal('40.31'), Decimal('540570.80')),
 (datetime.datetime(1997, 8, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('47287.67'), Decimal('51020.86'), Decimal('-7.32'), Decimal('587858.47')),
 (datetime.datetime(1997, 9, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('55629.24'), Decimal('47287.67'), Decimal('17.64'), Decimal('643487.71')),
 (datetime.datetime(1997, 10, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000))), Decimal('66749.23'), Decimal('55629.24'), Decimal('19.99'), Decimal('710236.94')),
 (datetime.datetime(1997, 11, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('43533.81'), Decimal('66749.23'), Decimal('-34.78'), Decimal('753770.74')),
 (datetime.datetime(1997, 12, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('71398.43'), Decimal('43533.81'), Decimal('64.01'), Decimal('825169.17')),
 (datetime.datetime(1998, 1, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('94222.11'), Decimal('71398.43'), Decimal('31.97'), Decimal('919391.28')),
 (datetime.datetime(1998, 2, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('99415.29'), Decimal('94222.11'), Decimal('5.51'), Decimal('1018806.57')),
 (datetime.datetime(1998, 3, 1, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=68400))), Decimal('104854.16'), Decimal('99415.29'), Decimal('5.47'), Decimal('112366

#### High Value Customers
- Identify customers with above-average order values
- Rank customers based on the quantity of above-average orders and highlight the top 5 customers
- Identify the top 20% of customers by total purchase volume.

In [63]:
%%sql

WITH customer_order_totals AS (
SELECT c.customer_id, c.order_id, SUM(o.unit_price * o.quantity * (1 - o.discount)) AS order_total
  FROM customer_orders AS c
  JOIN order_details AS o ON c.order_id = o.order_id
 GROUP BY c.customer_id, c.order_id
 ORDER BY c.customer_id, c.order_id),

order_categorization AS (
SELECT customer_id, order_id, order_total,
       CASE
        WHEN order_total > AVG(order_total) OVER w THEN 'Above Average'
        WHEN order_total <= AVG(order_total) OVER w THEN 'Average/Below Average'
       END AS order_category
  FROM customer_order_totals
WINDOW w AS (ORDER BY order_total RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)),

customer_ranking AS (
SELECT customer_id, COUNT(DISTINCT order_id) AS above_avg_order_count,
       ROW_NUMBER() OVER(ORDER BY COUNT(DISTINCT order_id) DESC) AS customer_rnk
  FROM order_categorization
 WHERE order_category = 'Above Average'
 GROUP BY customer_id)

SELECT customer_id, above_avg_order_count
  FROM customer_ranking
 WHERE customer_rnk <=5;

 * postgresql://postgres@localhost:5432/northwind
5 rows affected.


[('ERNSH', 26), ('SAVEA', 26), ('QUICK', 22), ('HUNGO', 11), ('RATTC', 10)]

In [144]:
%%sql

WITH customer_percentiles AS (
SELECT c.customer_id, c.company_name, SUM(od.quantity) AS purchase_volume, ROUND(PERCENT_RANK() OVER(ORDER BY SUM(od.quantity) DESC)::numeric,4) AS percentile
  FROM customers AS c
  JOIN orders AS o on c.customer_id = o.customer_id
  JOIN order_details AS od ON o.order_id = od.order_id
 GROUP BY c.customer_id, c.company_name)

SELECT customer_id, company_name, purchase_volume, percentile
  FROM customer_percentiles
 WHERE percentile <= 0.20

 * postgresql://postgres@localhost:5432/northwind
18 rows affected.


[('SAVEA', 'Save-a-lot Markets', 4958, Decimal('0.0000')),
 ('ERNSH', 'Ernst Handel', 4543, Decimal('0.0114')),
 ('QUICK', 'QUICK-Stop', 3961, Decimal('0.0227')),
 ('HUNGO', 'Hungry Owl All-Night Grocers', 1684, Decimal('0.0341')),
 ('FRANK', 'Frankenversand', 1525, Decimal('0.0455')),
 ('RATTC', 'Rattlesnake Canyon Grocery', 1383, Decimal('0.0568')),
 ('FOLKO', 'Folk och fä HB', 1234, Decimal('0.0682')),
 ('HILAA', 'HILARION-Abastos', 1096, Decimal('0.0795')),
 ('SUPRD', 'Suprêmes délices', 1072, Decimal('0.0909')),
 ('WHITC', 'White Clover Markets', 1063, Decimal('0.1023')),
 ('QUEEN', 'Queen Cozinha', 1031, Decimal('0.1136')),
 ('BERGS', 'Berglunds snabbköp', 1001, Decimal('0.1250')),
 ('BONAP', "Bon app'", 980, Decimal('0.1364')),
 ('LINOD', 'LINO-Delicateses', 970, Decimal('0.1477')),
 ('MEREP', 'Mère Paillarde', 966, Decimal('0.1591')),
 ('BOTTM', 'Bottom-Dollar Markets', 956, Decimal('0.1705')),
 ('KOENE', 'Königlich Essen', 903, Decimal('0.1818')),
 ('HANAR', 'Hanari Carnes', 839, Decimal('0.1932'))]

#### Product and Category Performance
- Assess the sales composition across different product categories by finding the percent of total sales for each
- Drill down into each group to determine the top three items sold in each product category
- Analyze the cumulative sales for each product category over a specified period to understand each category's growth.
- Identify products that perform exceptionally well in terms of sales. Specifically, products whose total sales exceed the 75th percentile.

In [148]:
%%sql

WITH aggregate_sales AS (
SELECT p.category_id, c.category_name, ROUND(SUM(unit_price * quantity * (1 - discount))::numeric,2) AS total_sales
  FROM product_quantity_orders AS p
  JOIN categories AS c ON p.category_id = c.category_id
 GROUP BY p.category_id, c.category_name
 ORDER BY total_sales DESC)

SELECT category_id, category_name, total_sales, ROUND(total_sales / (SELECT SUM(total_sales) FROM aggregate_sales) * 100.0,2) AS perc_of_total,
       SUM(total_sales) OVER(ORDER BY total_sales DESC) AS running_total,
       SUM(ROUND(total_sales / (SELECT SUM(total_sales) FROM aggregate_sales) * 100.0,2)) OVER(ORDER BY total_sales DESC) AS running_percent
  FROM aggregate_sales
 ORDER BY total_sales DESC;

 * postgresql://postgres@localhost:5432/northwind
8 rows affected.


[(1, 'Beverages', Decimal('288694.65'), Decimal('21.33'), Decimal('288694.65'), Decimal('21.33')),
 (4, 'Dairy Products', Decimal('251147.60'), Decimal('18.56'), Decimal('539842.25'), Decimal('39.89')),
 (3, 'Confections', Decimal('179863.65'), Decimal('13.29'), Decimal('719705.90'), Decimal('53.18')),
 (6, 'Meat/Poultry', Decimal('174622.55'), Decimal('12.90'), Decimal('894328.45'), Decimal('66.08')),
 (8, 'Seafood', Decimal('137989.31'), Decimal('10.20'), Decimal('1032317.76'), Decimal('76.28')),
 (2, 'Condiments', Decimal('113692.19'), Decimal('8.40'), Decimal('1146009.95'), Decimal('84.68')),
 (7, 'Produce', Decimal('105745.70'), Decimal('7.81'), Decimal('1251755.65'), Decimal('92.49')),
 (5, 'Grains/Cereals', Decimal('101646.94'), Decimal('7.51'), Decimal('1353402.59'), Decimal('100.00'))]

The most popular product categories are beverages, dairy products and confections which make up 53.18% of all sales ($719,705.90)

In [182]:
%%sql

WITH sales_by_product AS (
SELECT c.category_name, p.product_name, ROUND(SUM(unit_price * quantity * (1 - discount))::numeric,2) AS total_sales,
       ROW_NUMBER() OVER(PARTITION BY c.category_name ORDER BY SUM(unit_price * quantity * (1 - discount))::numeric DESC) AS product_ranking
  FROM product_quantity_orders AS p
  JOIN categories AS c ON p.category_id = c.category_id
 GROUP BY 1,2)

SELECT category_name, product_name, total_sales
  FROM sales_by_product
 WHERE product_ranking <= 3

 * postgresql://postgres@localhost:5432/northwind
24 rows affected.


[('Beverages', 'Côte de Blaye', Decimal('153897.17')),
 ('Beverages', 'Ipoh Coffee', Decimal('25109.10')),
 ('Beverages', 'Chang', Decimal('17719.40')),
 ('Condiments', 'Vegie-spread', Decimal('18343.62')),
 ('Condiments', "Sirop d'érable", Decimal('15022.35')),
 ('Condiments', 'Louisiana Fiery Hot Pepper Sauce', Decimal('14893.93')),
 ('Confections', 'Tarte au sucre', Decimal('50737.09')),
 ('Confections', "Sir Rodney's Marmalade", Decimal('24199.56')),
 ('Confections', 'Gumbär Gummibärchen', Decimal('21662.69')),
 ('Dairy Products', 'Raclette Courdavault', Decimal('76683.75')),
 ('Dairy Products', 'Camembert Pierrot', Decimal('49877.32')),
 ('Dairy Products', 'Mozzarella di Giovanni', Decimal('27086.58')),
 ('Grains/Cereals', 'Gnocchi di nonna Alice', Decimal('45351.10')),
 ('Grains/Cereals', 'Wimmers gute Semmelknödel', Decimal('23487.47')),
 ('Grains/Cereals', 'Singaporean Hokkien Fried Mee', Decimal('8986.60')),
 ('Meat/Poultry', 'Thüringer Rostbratwurst', Decimal('84783.77')),
 ('Meat/Poultry', 'Alice Mutton', Decimal('35105.85')),
 ('Meat/Poultry', 'Perth Pasties', Decimal('22623.80')),
 ('Produce', 'Manjimup Dried Apples', Decimal('43846.90')),
 ('Produce', 'Rössle Sauerkraut', Decimal('27936.84')),
 ('Produce', "Uncle Bob's Organic Dried Pears", Decimal('22453.50')),
 ('Seafood', 'Carnarvon Tigers', Decimal('30728.12')),
 ('Seafood', 'Ikura', Decimal('21653.50')),
 ('Seafood', 'Boston Crab Meat', Decimal('19055.04'))]

In [173]:
%%sql

WITH monthly_sales_by_category AS (
SELECT TO_CHAR(DATE_TRUNC('month', order_date),'YYYY-MM') AS order_month, c.category_name, ROUND(SUM(p.unit_price * od.quantity * (1 - od.discount))::numeric,2) AS total_sales
  FROM categories AS c
  JOIN products AS p ON c.category_id = p.category_id
  JOIN order_details AS od ON p.product_id = od.product_id
  JOIN orders AS o ON od.order_id = o.order_id
 GROUP BY c.category_name, order_month
 ORDER BY order_month)

SELECT order_month, category_name, total_sales, LAG(total_sales) OVER w1 AS previous_month_sales, ROUND((total_sales - LAG(total_sales) OVER w1) / LAG(total_sales) OVER w1 * 100.0,2) AS growth_rate
  FROM monthly_sales_by_category
WINDOW w1 AS (PARTITION BY category_name ORDER BY order_month)

 * postgresql://postgres@localhost:5432/northwind
184 rows affected.


[('1996-07', 'Beverages', Decimal('3978.12'), None, None),
 ('1996-08', 'Beverages', Decimal('6083.60'), Decimal('3978.12'), Decimal('52.93')),
 ('1996-09', 'Beverages', Decimal('6360.50'), Decimal('6083.60'), Decimal('4.55')),
 ('1996-10', 'Beverages', Decimal('10234.20'), Decimal('6360.50'), Decimal('60.90')),
 ('1996-11', 'Beverages', Decimal('21452.57'), Decimal('10234.20'), Decimal('109.62')),
 ('1996-12', 'Beverages', Decimal('11789.75'), Decimal('21452.57'), Decimal('-45.04')),
 ('1997-01', 'Beverages', Decimal('27380.20'), Decimal('11789.75'), Decimal('132.24')),
 ('1997-02', 'Beverages', Decimal('3557.30'), Decimal('27380.20'), Decimal('-87.01')),
 ('1997-03', 'Beverages', Decimal('13296.10'), Decimal('3557.30'), Decimal('273.77')),
 ('1997-04', 'Beverages', Decimal('7074.35'), Decimal('13296.10'), Decimal('-46.79')),
 ('1997-05', 'Beverages', Decimal('15422.25'), Decimal('7074.35'), Decimal('118.00')),
 ('1997-06', 'Beverages', Decimal('3485.42'), Decimal('15422.25'), Decimal('-77.40')),
 ('1997-07', 'Beverages', Decimal('7889.22'), Decimal('3485.42'), Decimal('126.35')),
 ('1997-08', 'Beverages', Decimal('5836.92'), Decimal('7889.22'), Decimal('-26.01')),
 ('1997-09', 'Beverages', Decimal('5726.70'), Decimal('5836.92'), Decimal('-1.89')),
 ('1997-10', 'Beverages', Decimal('8374.90'), Decimal('5726.70'), Decimal('46.24')),
 ('1997-11', 'Beverages', Decimal('3851.00'), Decimal('8374.90'), Decimal('-54.02')),
 ('1997-12', 'Beverages', Decimal('10876.65'), Decimal('3851.00'), Decimal('182.44')),
 ('1998-01', 'Beverages', Decimal('27245.40'), Decimal('10876.65'), Decimal('150.49')),
 ('1998-02', 'Beverages', Decimal('34599.15'), Decimal('27245.40'), Decimal('26.99')),
 ('1998-03', 'Beverages', Decimal('27761.57'), Decimal('34599.15'), Decimal('-19.76')),
 ('1998-04', 'Beverages', Decimal('22362.05'), Decimal('27761.57'), Decimal('-19.45')),
 ('1998-05', 'Beverages', Decimal('4056.70'), Decimal('22362.05'), Decimal('-81.86')),
 ('1996-07', 'Condiments', Decimal('2355.55'), None, None),
 ('1996-08', 'Condiments', Decimal('2827.47'), Decimal('2355.55'), Decimal('20.03')),
 ('1996-09', 'Condiments', Decimal('2268.35'), Decimal('2827.47'), Decimal('-19.77')),
 ('1996-10', 'Condiments', Decimal('5146.60'), Decimal('2268.35'), Decimal('126.89')),
 ('1996-11', 'Condiments', Decimal('7869.74'), Decimal('5146.60'), Decimal('52.91')),
 ('1996-12', 'Condiments', Decimal('1876.26'), Decimal('7869.74'), Decimal('-76.16')),
 ('1997-01', 'Condiments', Decimal('6570.78'), Decimal('1876.26'), Decimal('250.21')),
 ('1997-02', 'Condiments', Decimal('7671.21'), Decimal('6570.78'), Decimal('16.75')),
 ('1997-03', 'Condiments', Decimal('2057.50'), Decimal('7671.21'), Decimal('-73.18')),
 ('1997-04', 'Condiments', Decimal('5681.87'), Decimal('2057.50'), Decimal('176.15')),
 ('1997-05', 'Condiments', Decimal('5453.02'), Decimal('5681.87'), Decimal('-4.03')),
 ('1997-06', 'Condiments', Decimal('1830.28'), Decimal('5453.02'), Decimal('-66.44')),
 ('1997-07', 'Condiments', Decimal('5460.46'), Decimal('1830.28'), Decimal('198.34')),
 ('1997-08', 'Condiments', Decimal('4220.02'), Decimal('5460.46'), Decimal('-22.72')),
 ('1997-09', 'Condiments', Decimal('3450.18'), Decimal('4220.02'), Decimal('-18.24')),
 ('1997-10', 'Condiments', Decimal('6565.91'), Decimal('3450.18'), Decimal('90.31')),
 ('1997-11', 'Condiments', Decimal('3784.67'), Decimal('6565.91'), Decimal('-42.36')),
 ('1997-12', 'Condiments', Decimal('5824.20'), Decimal('3784.67'), Decimal('53.89')),
 ('1998-01', 'Condiments', Decimal('4737.90'), Decimal('5824.20'), Decimal('-18.65')),
 ('1998-02', 'Condiments', Decimal('6293.97'), Decimal('4737.90'), Decimal('32.84')),
 ('1998-03', 'Condiments', Decimal('10773.27'), Decimal('6293.97'), Decimal('71.17')),
 ('1998-04', 'Condiments', Decimal('10087.08'), Decimal('10773.27'), Decimal('-6.37')),
 ('1998-05', 'Condiments', Decimal('885.90'), Decimal('10087.08'), Decimal('-91.22')),
 ('1996-07', 'Confections', Decimal('7226.13'), None, None),
 ('19

In [181]:
%%sql

WITH product_percentiles AS (
SELECT c.category_name, p.product_name, ROUND(SUM(unit_price * quantity * (1 - discount))::numeric,2) AS total_sales,
       ROUND(PERCENT_RANK() OVER(ORDER BY SUM(unit_price * quantity * (1 - discount)))::numeric,4) AS product_percentile
  FROM product_quantity_orders AS p
  JOIN categories AS c ON p.category_id = c.category_id
 GROUP BY 1,2)

SELECT product_name, total_sales
  FROM product_percentiles
 WHERE product_percentile <= 0.25

 * postgresql://postgres@localhost:5432/northwind
20 rows affected.


[('Chocolade', Decimal('1554.86')),
 ('Genen Shouyu', Decimal('1561.95')),
 ('Geitost', Decimal('1818.00')),
 ('Laughing Lumberjack Lager', Decimal('2407.30')),
 ('Longlife Tofu', Decimal('2812.50')),
 ('Gravad lax', Decimal('2891.20')),
 ('Aniseed Syrup', Decimal('3240.00')),
 ('Filo Mix', Decimal('3341.45')),
 ('Valkoinen suklaa', Decimal('3741.56')),
 ('Louisiana Hot Spiced Okra', Decimal('3893.00')),
 ('NuNuCa Nuß-Nougat-Creme', Decimal('4091.50')),
 ('Röd Kaviar', Decimal('4174.20')),
 ('Zaanse koeken', Decimal('4186.65')),
 ('Rogede sild', Decimal('4412.27')),
 ('Guaraná Fantástica', Decimal('4775.85')),
 ('Tunnbröd', Decimal('4981.50')),
 ('Konbu', Decimal('5069.40')),
 ('Tourtière', Decimal('5214.63')),
 ("Chef Anton's Gumbo Mix", Decimal('5841.36')),
 ('Escargots de Bourgogne', Decimal('6226.17'))]